src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -220,42 +219,34 @@ public override IOperation VisitConditional(IConditionalOperation operation, obj
                 // result - capture
 
                 SpillEvalStack();
-                int captureId = _availableCaptureId++;
+                int captureId = captureIdForResult ?? _availableCaptureId++;
 
                 BasicBlock whenFalse = null;
                 VisitConditionalBranch(operation.Condition, ref whenFalse, sense: false);
 
-                AddStatement(new SimpleAssignmentExpression(new FlowCapture(captureId, operation.Syntax,
-                                                                            isInitialization: true, operation.Type,
-                                                                            default(Optional<object>)),
-                                                            operation.IsRef,
-                                                            Visit(operation.WhenTrue),
-                                                            semanticModel: null,
-                                                            operation.WhenTrue.Syntax,
-                                                            operation.Type,
-                                                            default(Optional<object>),
-                                                            isImplicit: true));
+                IOperation whenTrueValue = Visit(operation.WhenTrue, captureId);
+                if (whenTrueValue.Kind != OperationKind.FlowCaptureReference ||
+                    captureId != ((IFlowCaptureReferenceOperation)whenTrueValue).Id)
```

Comment:
There are several instances of this check. Consider extracting a `IsFlowCaptureReference(IOperation op, int id)` helper method.

Reply:
> There are several instances of this check. Consider extracting an IsFlowCaptureReference(IOperation op, int id) helper method.

Good idea. Will do that in the next PR.

---
In reply to: [163282938](https://github.com/dotnet/roslyn/pull/24358#discussion_r163282938) [](ancestors = 163282938)